service/src/main/java/uk/nhs/adaptors/gpc/consumer/filters/SdsFilter.java

```diff
@@ -0,0 +1,162 @@
+            prepareLookupUri(address, serverHttpRequest.getPath()).ifPresent(uri -> exchange.getAttributes()
+                .put(ServerWebExchangeUtils.GATEWAY_REQUEST_URL_ATTR, uri));
+        } else {
+            LOGGER.error("SDS filter request was unsuccessful.");
+        }
+    }
+
+    private Optional<SdsClient.SdsResponseData> performRequestAccordingToInteractionId(String interactionId,
+            ServerHttpRequest serverHttpRequest) {
+        if (isRequestFunctionAvailable(interactionId)) {
+            Optional<String> organisation = extractOrganisation(serverHttpRequest
+                .getPath());
+
+            if (organisation.isPresent()) {
+                LOGGER.info("Performing request with organisation \"{}\" and NHS service endpoint id \"{}\"",
+                    organisation.get(), interactionId);
+                return sdsRequestFunctions.get(interactionId)
+                    .apply(organisation.get());
+            }
+        }
+        return Optional.empty();
+    }
+
+    private boolean isRequestFunctionAvailable(String interactionId) {
+        return sdsRequestFunctions.keySet()
+            .stream()
+            .anyMatch(key -> key.equals(interactionId));
+    }
+
+    private Optional<String> extractOrganisation(RequestPath requestPath) {
+        String path = requestPath.toString();
+        if (path.contains(SLASH)) {
+            String[] pathElements = path.split(SLASH);
+
+            if (pathElements.length > 1) {
+                return Optional.of(pathElements[1]);
+            }
+        }
+        return Optional.empty();
+    }
+
+    private Optional<String> extractInteractionId(HttpHeaders httpHeaders) {
+        if (httpHeaders.containsKey(SSP_INTERACTION_ID)) {
+            List<String> interactionIds = httpHeaders.get(SSP_INTERACTION_ID);
+
+            if (!CollectionUtils.isEmpty(interactionIds)) {
+                return Optional.of(interactionIds.get(0));
+            }
+        }
+        return Optional.empty();
+    }
+
+    private Optional<URI> prepareLookupUri(String address, RequestPath requestPath) {
+        String path = requestPath.value();
+        if (StringUtils.isNotBlank(address)) {
+            try {
+                String uri = address + extractUriSuffix(path);
+                return Optional.of(new URI(uri));
+            } catch (URISyntaxException e) {
+                LOGGER.error("Invalid address in SDS: " + address);
+                return Optional.empty();
+            }
+        }
+        return Optional.empty();
+    }
+
+    private String extractUriSuffix(String path) {
+        if (path.contains(DOUBLE_SEPARATOR)) {
+            return path.split(DOUBLE_SEPARATOR)[1];
+        } else if (path.contains(FHIR_SEPARATOR)) {
+            return path.split(FHIR_SEPARATOR)[1];
+        } else if (path.contains(DOCUMENTS_SEPARATOR)) {
+            return path.split(DOCUMENTS_SEPARATOR)[1];
+        }
+        return StringUtils.EMPTY;
+    }
```

Comment:
This could be simplified. We know the first five elements are always the base path and the rest is the FHIR path. Tests don't pass with this code, needs a bit of work to get the path right I think.

```suggestion
private Optional<URI> prepareLookupUri(String address, RequestPath requestPath) {
        String fhirPath = requestPath.subPath(5).toString();
        try {
            String uri = address + fhirPath;
            return Optional.of(new URI(uri));
        } catch (URISyntaxException e) {
            throw new RuntimeException("Invalid address in SDS: " + address);
        }
    }
```